### PR TITLE
By default, prevent `BugChecker`s from introducing new dependencies

### PR DESCRIPTION
**Pull request by Stephan202**
As-is, this is especially relevant for users who do not use Guava or the New Relic Agent API, as mentioned [here](https://www.reddit.com/r/java/comments/y2u8tv/comment/is50d1u/?utm_source=share&utm_medium=web2x&context=3). Note that most `BugChecker`s don't need to be modified, since any third-party references in their suggestions must already be present if the check is to trigger at all.

Suggested commit message:
```
By default, prevent `BugChecker`s from introducing new dependencies (#308)
```

We should implement something similar for Refaster, but that will look quite different (and is out of scope for this PR).

### CODE REVIEW

**Review by rickie — COMMENTED** [review]
Added a commit. 

First round of review 👀:smile:.

**Review by Stephan202 — COMMENTED** [review]
Tnx for the typo fixes :)

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/util/ThirdPartyLibrary.java
Comment (Stephan202) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/util/ThirdPartyLibrary.java:63]: Note to self: ponder a better name. I didn't select `isSupported` because it doesn't seem to jive with `IGNORE_CLASSPATH_COMPAT_FLAG`, but perhaps we should use that name anyway. Or perhaps something else?

**Review by oxkitsune — APPROVED** [review]
Cool stuff! Left some questions and opinions :smile: 

Would be interesting to see whether we could make a bug checker that enforces `ThirdPartyLibrary` usage in bug checkers :exploding_head:

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/util/ThirdPartyLibrary.java
Comment (oxkitsune) [reply error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/util/ThirdPartyLibrary.java:63]: I think `canUse` is fine, especially if we plan to add more fine-grained control.
In this case I think `canUse` makes more sense than `isSupported` (or `isAvailable`) as the library might be available on the classpath, but like the XXX mentions new usages could be undesirable.

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/util/ThirdPartyLibrary.java
Comment (oxkitsune) [reply error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/util/ThirdPartyLibrary.java:77]: I prefer `className` as it's more intuitive than `fqcn` :+1:

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/util/ThirdPartyLibrary.java
Comment (rickie) [reply error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/util/ThirdPartyLibrary.java:63]: Yeah I was also thinking about this. 

`isAllowedToUse` (is allowed might a bit more "accurate" as `shouldIgnoreClasspath` is part of that check, although it is very similar to `canUse`)?

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/util/ThirdPartyLibrary.java
Comment (rickie) [reply error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/util/ThirdPartyLibrary.java:97]: Because the `getBoolean` expects a `Boolean` and otherwise there would be an implicit cast from the boxed `false` variant.

**Review by rickie — APPROVED** [review]
Some final notes, PR looks nice!

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/ScheduledTransactionTrace.java
Comment (Stephan202) [reply error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/ScheduledTransactionTrace.java]: I see a +1, so will do.

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/StaticImport.java
Comment (Stephan202) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/StaticImport.java:106]: This doesn't cause the enum values to be statically imported. Need to investigate before merging this PR.

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/StaticImport.java
Comment (Stephan202) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/StaticImport.java:194]: ... but `ThirdPartyLibrary.values()`, _would_ be statically imported, which is not something we want.

**Review by Stephan202 — COMMENTED** [review]
Rebased and added a commit. We could also move the `StaticImport` changes and investigation to a separate PR. (If we do want to keep it here, then eventually the suggested commit message should be updated, too.)

**Comment by rickie** [comment]
Note that I mention in [this comment](https://github.com/PicnicSupermarket/error-prone-support/pull/308#discussion_r1003514595) that I want to add it to the `STATIC_IMPORT_EXEMPTED_MEMBERS` because I _don't_ like statically importing this one. Primarily as it would mean loosing some context. 

> ... but ThirdPartyLibrary.values(), would be statically imported, which is not something we want.

That is not true right? It would be prevented in the `isCandidate` method because it sees that `values` is in `STATIC_IMPORT_EXEMPTED_IDENTIFIERS`. 

So I want to propose that I add a commit to move it. Which means that we defer looking into the issue with not importing 🤔. WDYT?

**Comment by rickie** [comment]
I see that we currently only allow to exempt specific _members_ using `STATIC_IMPORT_EXEMPTED_MEMBERS`, which we could do as there are only 4 types in the `enum`. Nevertheless, let's defer it all together and investigate the issue + introduce support for easily adding a class to disallow all members thereof.

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/util/ThirdPartyLibrary.java
Comment (rickie) [reply error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/util/ThirdPartyLibrary.java]: Read the Javadocs and goal of the class again.

Have three other options: `isUsageAllowed`, `isReferencingAllowed`, `isReferenceAllowed`. 

To me, the `ThirdPartyLibrary#isIntroductionAllowed` sounds a bit like the question: Is it okay if we introduce this third party library (at all)? The name of the method does not 100% match with the Javadoc of the class, as it mentions "introducing **references**". So I'd say that `isReferencingAllowed` could make the whole intent of the class even clearer (and is for that reason the one I like most). WDYT?

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/StaticImport.java
Comment (Stephan202) [reply error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/StaticImport.java]: This should also be reverted.

**Review by Stephan202 — COMMENTED** [review]
Rebased and added a commit. As discussed offline: will merge once built.

> Note that I mention in [this comment](https://github.com/PicnicSupermarket/error-prone-support/pull/308#discussion_r1003514595) that I want to add it to the `STATIC_IMPORT_EXEMPTED_MEMBERS` because I _don't_ like statically importing this one. Primarily as it would mean loosing some context.

Okay, I misread that (but find it rather surprising; I don't see any issue with statically importing these, as long as it'd be done consistently :shrug:).
 
> > ... but ThirdPartyLibrary.values(), would be statically imported, which is not something we want.
> 
> This is not true right? It would be prevented in the `isCandidate` method because it sees that `values` is in `STATIC_IMPORT_EXEMPTED_IDENTIFIERS`.

Yes, because I [added](https://github.com/PicnicSupermarket/error-prone-support/pull/308/commits/13c862beaa90f40b8d59af949bfd209ab605b10e#diff-e0df674a0d2abd3814f43a388eb9244a2becdb9e634ec98f9081d3e980a94381R194) `values` in the context of this PR after making this observation...

**Comment by rickie** [comment]
> I don't see any issue with statically importing these

I think it is a bit less clear 🤷🏻.

> Yes, because I [added](https://github.com/PicnicSupermarket/error-prone-support/pull/308/commits/13c862beaa90f40b8d59af949bfd209ab605b10e#diff-e0df674a0d2abd3814f43a388eb9244a2becdb9e634ec98f9081d3e980a94381R194) values in the context of this PR after making this observation...

Ahh I missed that change 😂, that's why I didn't revert it haha. Thanks for fixing that 😉.

**Pull request merged**
